test/test_torch.py

```diff
@@ -17955,92 +17955,155 @@ def test_case_info(fn_name, config):
-            self.assertEqual(res2, torch.addbmm(1, res2, 0, b1, b2))
+                UserWarning, f"This overload of {func} is deprecated"):
+            self.assertEqual(res2, getattr(torch, func)(1, res2, 0, b1, b2))
 
-        res4 = torch.addbmm(res2, b1, b2, beta=1, alpha=.5)
-        self.assertEqual(res4, res.sum(0, False) * 3),
+        res4 = getattr(torch, func)(res2, b1, b2, beta=1, alpha=.5)
+        self.assertEqual(res4, ref * 3),
 
-        res5 = torch.addbmm(res2, b1, b2, beta=0, alpha=1)
-        self.assertEqual(res5, res.sum(0, False))
+        nan = torch.full_like(res2, math.nan)
+        res5 = getattr(torch, func)(nan, b1, b2, beta=0, alpha=1)
+        self.assertEqual(res5, ref)
 
-        res6 = torch.addbmm(res2, b1, b2, beta=.1, alpha=.5)
-        self.assertEqual(res6, res2 * .1 + .5 * res.sum(0)),
-
-    @onlyCPU
-    @dtypes(*(torch.testing.get_all_complex_dtypes() + [torch.float, torch.double]))
-    def test_baddbmm(self, device, dtype):
-        num_batches = 10
-        M, N, O = 12, 8, 5
-        b1 = torch.randn(num_batches, M, N, dtype=dtype, device=device)
-        b2 = torch.randn(num_batches, N, O, dtype=dtype, device=device)
-        res = torch.bmm(b1, b2)
-        res2 = torch.tensor((), dtype=dtype, device=device).resize_as_(res).zero_()
-        res3 = torch.tensor((), dtype=dtype, device=device).resize_as_(res).zero_()
+        if b1.is_complex():
+            res6 = getattr(torch, func)(res2, b1, b2, beta=.1j, alpha=.5j)
+            self.assertEqual(res6, res2 * .1j + .5j * ref)
+        else:
+            res6 = getattr(torch, func)(res2, b1, b2, beta=.1, alpha=.5)
+            self.assertEqual(res6, res2 * .1 + .5 * ref)
 
-        res2.baddbmm_(b1, b2)
-        self.assertEqual(res2, res)
-        res3.copy_(res2)
+        res7 = torch.full_like(res2, math.nan)
+        getattr(torch, func)(nan, b1, b2, beta=0, out=res7)
+        self.assertEqual(res7, ref)
 
-        with self.maybeWarnsRegex(
-                UserWarning, "This overload of baddbmm_ is deprecated"):
-            res2.baddbmm_(1, b1, b2)
-        self.assertEqual(res2, res * 2)
-        res3.baddbmm_(b1, b2, beta=1)
-        self.assertEqual(res3, res2)
+    @precisionOverride({torch.half: 0.05, torch.bfloat16: 0.05})
+    @onlyOnCPUAndCUDA
+    @dtypes(*torch.testing.get_all_fp_dtypes(), *torch.testing.get_all_complex_dtypes())
+    @tf32_on_and_off(0.05)
+    def test_addbmm(self, device, dtype):
+        num_batches = 2
+        M, N, O = 2, 3, 4
 
-        with self.maybeWarnsRegex(
-                UserWarning, "This overload of baddbmm_ is deprecated"):
-            res2.baddbmm_(1, .5, b1, b2)
-        self.assertEqual(res2, res * 2.5)
-        res3.baddbmm_(b1, b2, beta=1, alpha=.5)
-        self.assertEqual(res3, res2)
+        if self.device_type == 'cpu':
+            is_supported = True
+            if dtype == torch.bfloat16:
+                self.precision = 1  # 43 vs 43.75
```

Comment:
Reviewer's note: It'd be nicer if we supported per-dtype precision overrides (also atol and rtol overrides)